File: C++/010. Regular Expression Matching.cpp
class Solution {
public:
    bool isMatch(string s, string p) 
    {
        int m = s.length();
        int n = p.length();
        vector<vector<bool>> dp = vector<vector<bool>>(m+1, vector<bool>(n+1, false));
        dp[0][0] = true;
        for(int j = 2; j <= n; j += 2)
        {
            dp[0][j] = (dp[0][j-2] && p[j-1] == '*');
        }
        for (int i = 1; i <= m; i++)
        {
            for (int j = 1; j <= n; j++)
            {
                char c1 = s[i-1];
                char c2 = p[j-1];
                if (c2 != '*' && c2 != '.')
                {
                    dp[i][j] = dp[i-1][j-1] && (c1 == c2);
                }
                else if (c2 == '.')
                {
                    dp[i][j] = dp[i-1][j-1];
                }
                else if (j > 1)
                {
                    if (dp[i][j-2] || dp[i][j-1])
                    {
                        dp[i][j] = true;
                    }
                    else if (dp[i-1][j] && (s[i-1] == p[j-2] || p[j-2] == '.') ) 
                    {
                        dp[i][j] = true;
                    }
                }
            }
        }
        return dp[m][n];
    }
};

File: python/010. Regular Expression Matching.py
class Solution(object):
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        n = len(s)
        m = len(p)
        dp = [[False] * (m+1) for i in range(n+1)]
        dp[0][0] = True
        i = 1
        for j in range(2,m+1,2):
            dp[0][j] = dp[0][j-2] and p[j-1] == "*"
        for i in range(1,n+1):
            for j in range(1,m+1):
                if p[j-1] == "*":
                    if j > 1:
                        if dp[i][j-2]:
                            dp[i][j] = True
                        elif dp[i-1][j] and (s[i-1] == p[j-2] or p[j-2] == '.'):
                            dp[i][j] = True
                        
                elif p[j-1] == ".":
                    dp[i][j] = dp[i-1][j-1]
                else:
                    dp[i][j] = (p[j-1] == s[i-1] and dp[i-1][j-1])
        return dp[n][m]
